src/components/ripple/ripple.ts

```diff
@@ -0,0 +1,288 @@
+      this.end(event.pageX, event.pageY, isKeyEvent);
+    }
+  }
+
+  mouseLeave(event: MouseEvent) {
+    // We can always fade out the background here; It's a no-op if it was already inactive.
+    this._rippleManager.hideRippleBackground(this._rippleBackground);
+  }
+
+  // TODO: Reactivate the background div if the user drags out and back in.
+}
+
+/**
+ * Helper service that performs DOM manipulations. Not intended to be used outside this module.
+ */
+class MdInkRippleManager {
+  constructor(private _renderer: Renderer) {}
+
+  createForegroundRipple(
+      container: Element,
+      rippleOriginLeft: number,
+      rippleOriginTop: number,
+      color: string,
+      centered: boolean,
+      speedFactor: number,
+      transitionEndCallback: (r: Ripple, e: TransitionEvent) => void) {
+    const parentRect = container.getBoundingClientRect();
+    // Create a foreground ripple div with the size and position of the fully expanded ripple.
+    // When the div is created, it's given a transform style that causes the ripple to be displayed
+    // small and centered on the event location (or the center of the bounding rect if the centered
+    // input is true). Removing that transform causes the ripple to animate to its natural size.
+    const startX = centered ? (parentRect.left + parentRect.width / 2) : rippleOriginLeft;
+    const startY = centered ? (parentRect.top + parentRect.height / 2) : rippleOriginTop;
+    const offsetX = startX - parentRect.left;
+    const offsetY = startY - parentRect.top;
+    const maxRadius = distanceToFurthestCorner(startX, startY, parentRect);
+
+    const rippleDiv = this._renderer.createElement(container, 'div', null);
+    this._renderer.setElementClass(rippleDiv, 'md-ripple-foreground', true);
+    this._renderer.setElementStyle(rippleDiv, 'left', (offsetX - maxRadius) + 'px');
+    this._renderer.setElementStyle(rippleDiv, 'top', (offsetY - maxRadius) + 'px');
+    this._renderer.setElementStyle(rippleDiv, 'width', (2 * maxRadius) + 'px');
+    this._renderer.setElementStyle(rippleDiv, 'height', (2 * maxRadius) + 'px');
+    // If color input is not set, this will default to the background color defined in CSS.
+    this._renderer.setElementStyle(rippleDiv, 'background-color', color);
+
+    const translateX = offsetX - parentRect.width / 2;
+    const translateY = offsetY - parentRect.height / 2;
+    this._renderer.setElementStyle(rippleDiv,
+        'transform', `scale(0.01) translate(${translateX}px, ${translateY}px)`);
+
+    const fadeInSeconds = (1 / (speedFactor || 1)) * Math.max(
+        MIN_RIPPLE_FILL_TIME_SECONDS,
+        Math.min(MAX_RIPPLE_FILL_TIME_SECONDS, maxRadius / RIPPLE_SPEED_PX_PER_SECOND));
+    this._renderer.setElementStyle(rippleDiv, 'transition-duration', `${fadeInSeconds}s`);
+
+    // https://timtaubert.de/blog/2012/09/css-transitions-for-dynamically-created-dom-elements/
+    window.getComputedStyle(rippleDiv).opacity;
```

Comment:
I read through the linked article and think that it's now out of date (written in 2012).[ I tried doing this with `requestAnimationFrame` instead](https://jsbin.com/riyifacaqa/1/edit?css,js,output) and found that it works in Chrome, IE11, Edge, Firefox, and Safari.